superpixel_classification/SuperpixelClassification/SuperpixelClassificationBase.py

```diff
@@ -317,57 +350,69 @@ def createFeaturesForItem(self, gc, item, elem, featureFolderId, fileName, patch
             tsMask = large_image.open(maskPath)
 
             with h5py.File(filePath, 'w') as fptr:
-                for idx, _ in enumerate(elem['values']):
-                    prog.item_progress(item, 0.9 * idx / len(elem['values']))
-                    bbox = elem['user']['bbox'][idx * 4: idx * 4 + 4]
-                    # use masked superpixel
-                    patch = ts.getRegion(
-                        region=dict(
-                            left=int(bbox[0]), top=int(bbox[1]),
-                            right=int(bbox[2]), bottom=int(bbox[3])),
-                        output=dict(maxWidth=patchSize, maxHeight=patchSize),
-                        fill='#000',
-                        format=large_image.constants.TILE_FORMAT_NUMPY)[0]
-                    if patch.shape[2] in (2, 4):
-                        patch = patch[:, :, :-1]
-                    scale = 1
-                    try:
-                        scale = elem['transform']['matrix'][0][0]
-                    except Exception:
-                        pass
-                    mask = tsMask.getRegion(
-                        region=dict(
-                            left=int(bbox[0] / scale), top=int(bbox[1] / scale),
-                            right=int(bbox[2] / scale), bottom=int(bbox[3] / scale)),
-                        output=dict(maxWidth=patchSize, maxHeight=patchSize),
-                        fill='#000',
-                        format=large_image.constants.TILE_FORMAT_NUMPY)[0]
-                    if mask.shape[2] == 4:
-                        mask = mask[:, :, :-1]
-                    maskvals = [[val % 256, val // 256 % 256, val // 65536 % 256]
-                                for val in [idx * 2, idx * 2 + 1]]
-                    patch = patch.copy()
-                    patch[(mask != maskvals[0]).any(axis=-1) &
-                          (mask != maskvals[1]).any(axis=-1)] = [0, 0, 0]
-                    # TODO: ensure this is uint8
-                    if not ds:
-                        ds = fptr.create_dataset(
-                            'images', (1,) + patch.shape, maxshape=(None,) + patch.shape,
-                            dtype=patch.dtype, chunks=True)
-                    else:
-                        ds.resize((ds.shape[0] + 1,) + patch.shape)
-                    ds[ds.shape[0] - 1] = patch
-                    if time.time() - lastlog > 5:
-                        lastlog = time.time()
-                        print(ds.shape, len(elem['values']),
-                              '%5.3f' % (time.time() - starttime),
-                              '%5.3f' % ((len(elem['values']) - idx - 1) / (idx + 1) *
-                                         (time.time() - starttime)),
-                              item['name'])
+                batch_size = 1024  # TODO: Is this the best value?
```

Comment:
In other places we did some work to auto pick a batch size.  Is there a sensible way to do that?  Otherwise, what are the implications of different batch sizes here?

Reply:
The overhead per batch adds up to painfully slow if there are 100,000s-plus batches.  At the level of 100s-1000s of batches the runtime cost is pretty negligible.  If `batch_size` is too large then we run out of memory.  A value of 65536 used too much memory on my system.  The `batch_size` of 1024 is much faster than the old value of (effectively) 1, but could be a little faster.  The optimal value may be 2048 or 4096 or so.  But I think that we're close enough to defer that fine tuning to a future pull request.

Reply:
Sounds good.